Add commands for the device to run through the API

rossandrews opened the pull request:
Biggest change was adding a post request to the API that allows for commands to be added. Then I wrapped that in a get request that allows for one off command requests and outputs the results. /devices/{deviced_id}/runCommandCode{command_code}

Also did some refactoring:
-moved Result into the Command folder
-Took the command object out of Result to help with circular dependencies. 
-Changed command.name to command.code, makes more sense to me since it's possible to setup multiple commands using the same code with different triggers.

If you want to try it out go to <API-ip>/docs in your browser and it will show all the endpoints and allow you to make test calls.

Things I could work on next, let me know what you would like or if you have something else in mind:
1) Delete commands from device
2) Save configuration so it remains after a restart
3) Better structure for results
4) Home Assistant integration
5) Basic UI to schedule and run commands with

rossandrews commented on powermon/libs/apicoordinator.py:
```diff
@@ -48,29 +51,44 @@ def __init__(self, adhocTopic : str, announceTopic: str, enabled: bool, refreshI
-        if not self.last_run or time() - self.last_run > self.refreshInterval:
+        if not self.last_run or time() - self.last_run > self.refresh_interval:
             log.info("APICoordinator running")
             self.announce_device()  # QUESTION: what are we announcing here?
```
This is putting the information about the device on the topic "powermon/announce". This gets read by the API then it subscribes to the "powermon/{device_id}/results/#" topic to get results from commands being executed. I haven't tested it yet but this should allow the API to work with multiple devices sending results.

jblance commented on powermon/libs/apicoordinator.py:
```diff
@@ -12,32 +16,31 @@ class ApiCoordinator:
 
-        return cls(adhocTopic=adhocTopic, announceTopic=announceTopic, enabled=enabled, refreshInterval=refreshInterval, device=device, mqtt_broker=mqtt_broker)
+        adhoc_topic_format = "powermon/{device_id}/addcommand"
+        announce_topic = "powermon/announce"
```
these should probably be configurable

rossandrews replied:
I thought about that, it needs to be the same in the API so I put some of the config in the announce payload. It's not really clear to me where it should go though. Unless you have an idea where it should go I'll leave it hardcoded for a bit.

I could move it to the DTOs which are shared between the two.

jblance replied:
what about adding a section to the config file, was playing with it in a fork of your work (to see how it works)
eg:
```yaml
api:
  host: "0.0.0.0"
  port: 5000
  log_level: "info"
  announce_topic: powermon/announce
  adhoc_topic: powermon/adhoc
  refresh_interval: 60
```

jblance commented on powermon/ports/abstractport.py:
```diff
@@ -40,7 +41,7 @@ def disconnect(self) -> None:
 
     @abstractmethod
-    def send_and_receive(self, result) -> Result:
+    def send_and_receive(self, command: Command) -> Result:
```
this is much better logic :-)

jblance reviewed (commented):
Wow lots of improvements here - great work

rossandrews commented on powermon/dto/deviceDTO.py:
```diff
@@ -8,7 +8,7 @@
 # TODO: update
 class DeviceDTO(BaseModel):
-    identifier: Optional[str]
+    identifier: str
```
This is a data transfer object (DTO) that's a shared model between the API and the Powermon instance. This id is used to track which device results are coming from and for pulling data from the API. 

I'll rename it to device_id to be consistent with other naming conventions. I will also set it to a random number if none is provided in the config so it's optional in the configuration file.

jblance commented on api/app/statehandler.py:
```diff
@@ -49,7 +49,7 @@ def is_command_result_topic(self, topic: str) -> bool:
     def recieved_announcement(self, message) -> DeviceDTO:
         device = DeviceDTO.parse_raw(message)
-        deviceId : str = device.identifier 
+        deviceId : str = device.device_id 
```
there are more device.identifier references - eg hass formatter

rossandrews replied:
Good catch, searched the code base for ".identifier" without any hits so I think that was the last one.

rossandrews commented:
> can you make sure that the tests pass (and ideally have tests for new functionality)

I didn't realize there were powermon tests mixed in with the mpp-solar tests. Would you mind if I move them into a different folder?

jblance commented:
> > can you make sure that the tests pass (and ideally have tests for new functionality)
> 
> I didn't realize there were powermon tests mixed in with the mpp-solar tests. Would you mind if I move them into a different folder?

happy for the tests to be split and moved (they are somewhat limited so far, but I am trying to use the tests to reduce bugs as changes are made)

rossandrews commented:
> > > can you make sure that the tests pass (and ideally have tests for new functionality)
> > 
> > 
> > I didn't realize there were powermon tests mixed in with the mpp-solar tests. Would you mind if I move them into a different folder?
> 
> happy for the tests to be split and moved (they are somewhat limited so far, but I am trying to use the tests to reduce bugs as changes are made)

Unit tests are an excellent way to make code more stable, currently how the tests are written they are more like integration tests since the main command is called instead of a specific object. Fortunately I think the powermon code is in a good place to add some unit tests.

Are the tests I should move the three with the word powermon in it?

jblance commented:
> > > > can you make sure that the tests pass (and ideally have tests for new functionality)
> > > 
> > > 
> > > I didn't realize there were powermon tests mixed in with the mpp-solar tests. Would you mind if I move them into a different folder?
> > 
> > 
> > happy for the tests to be split and moved (they are somewhat limited so far, but I am trying to use the tests to reduce bugs as changes are made)
> 
> Unit tests are an excellent way to make code more stable, currently how the tests are written they are more like integration tests since the main command is called instead of a specific object. Fortunately I think the powermon code is in a good place to add some unit tests.
> 

the powermon ones are, there are some mppsolar ones that are more unittest-like, all in all the tests are patchy and usually targetted at areas that I have refactored

> Are the tests I should move the three with the word powermon in it?

yes + test_config_validation

rossandrews commented:
Moved the tests, got them running. Fixed one error and commented out an assert. Will work on converting that test to a unit test style so when they fail it's easy to find the cause.

jblance commented:
> Moved the tests, got them running. Fixed one error and commented out an assert. Will work on converting that test to a unit test style so when they fail it's easy to find the cause.

Hmm, but the test is a valid fail
There is a new `set_command_description(self, command_description):`, in the table formatter - where does this get called?

So running the below doesnt give the correct response (this is one of the reasons I am doing 'lazy' integration tests :-)

` $ powermon -1 --config '{"device": {"port":{"type":"test"}}, "commands": [{"command":"QPI", "outputs": [{"type": "screen", "format": "table"}]}]} ' `
```
-------------------------------------
Command: QPI - unknown command     
-------------------------------------
Parameter    Value   Unit       
protocol_id  PI30               
```

jblance commented on powermon/tests/test_powermon_formats.py:
```diff
@@ -110,7 +110,7 @@ def test_format_table(self):
             )
             # print(result.stdout)
-            self.assertEqual(result.stdout, expected)
+            #self.assertEqual(result.stdout, expected) #returning 'unknown command', not sure why and this type of test doesn't make it obvious. Will make soeme unittest style tests
```
need to fix the reason that this fails

rossandrews commented:
I've got a fix in the works, it's getting pretty big though. I'm creating a
CommandDefinition object so it's easier to track how the data is retrieved
from the protocols. This will also be useful for passing the information to
the API for setting up commands to run.

On Thu, Aug 31, 2023 at 2:23 PM John Blance ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In powermon/tests/test_powermon_formats.py
> <https://github.com/jblance/mpp-solar/pull/395#discussion_r1312301458>:
>
> > @@ -110,7 +110,7 @@ def test_format_table(self):
>                  text=True,
>              )
>              # print(result.stdout)
> -            self.assertEqual(result.stdout, expected)
> +            #self.assertEqual(result.stdout, expected) #returning 'unknown command', not sure why and this type of test doesn't make it obvious. Will make soeme unittest style tests
>
> need to fix the reason that this fails
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jblance/mpp-solar/pull/395#pullrequestreview-1605796878>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGYGGM6HJ37V6VQJHLS4TLXYD6DNANCNFSM6AAAAAA4AVIYEI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jblance commented:
> I've got a fix in the works, it's getting pretty big though. I'm creating a CommandDefinition object so it's easier to track how the data is retrieved from the protocols. This will also be useful for passing the information to the API for setting up commands to run.
> […](#)
> On Thu, Aug 31, 2023 at 2:23 PM John Blance ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In powermon/tests/test_powermon_formats.py <[#395 (comment)](https://github.com/jblance/mpp-solar/pull/395#discussion_r1312301458)>: > @@ -110,7 +110,7 @@ def test_format_table(self): text=True, ) # print(result.stdout) - self.assertEqual(result.stdout, expected) + #self.assertEqual(result.stdout, expected) #returning 'unknown command', not sure why and this type of test doesn't make it obvious. Will make soeme unittest style tests need to fix the reason that this fails — Reply to this email directly, view it on GitHub <[#395 (review)](https://github.com/jblance/mpp-solar/pull/395#pullrequestreview-1605796878)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAGYGGM6HJ37V6VQJHLS4TLXYD6DNANCNFSM6AAAAAA4AVIYEI> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Maybe revert the format first then we can commit this PR and then do the big fix

rossandrews commented:
Fixed the tests with a start to the new structure.

jblance commented:
think i will merge this soon (2-3 hours), to minimise variance (and so I can build on it)
let me know where you are focusing (and if there are any areas you want me to work or not work)

rossandrews commented:
It's late for me, I'll pick it up again tomorrow. I'm going to work on adding structure to the command definitions.

rossandrews commented:
Anything left before it's merged?
